docs/src/tutorial-nondeterministic-variables.md

```diff
@@ -41,26 +41,45 @@ The assertion we wrote in this harness was just an extra check we added to demon
 
 ## Custom nondeterministic types
 
-While `kani::any()` is the only method Kani provides to inject non-determinism into a proof harness, Kani only ships with implementations for a few types where we can guarantee safety.
-When you need nondeterministic variables of types that `kani::any()` cannot construct, you have two options:
-
-1. Implement the `kani::Arbitrary` trait for your type, so you can use `kani::any()`.
-2. Just write a function.
-
-The advantage of the first approach is that it's simple and conventional.
-It also means that in addition to being able to use `kani::any()` with your type, you can also use it with `Option<MyType>` (for example).
-
-The advantage of the second approach is that you're able to pass in parameters, like bounds on the size of the data structure.
+While `kani::any()` is the only method Kani provides to inject non-determinism into a proof harness, Kani only ships with implementations for a few `std` types where we can guarantee safety.
+When you need nondeterministic variables of types that don't have a `kani::any()` implementation available, you have the following options:
+
+1. Implement the `kani::Arbitrary` trait for your type, so you and downstream crates can use `kani::any()` with your type.
+2. Implement the [`bolero_engine::TypeGenerator` trait](https://docs.rs/bolero-engine/0.8.0/bolero_engine/trait.TypeGenerator.html).
+This will enable you and downstream crates to use Kani via [Bolero](https://camshaft.github.io/bolero/).
+3. Write a function that build an object from non-deterministic variables.
+
+We recommend the first approach for most cases.
+The first approach is simple and conventional. This option will also enable you to use it with parameterized types, such as `Option<MyType>` and arrays.
+Kani includes a derive macro that allows you to automatically derive `kani::Arbitrary` for structures and enumerations as long as all its fields also implement the `kani::Arbitrary` trait.
+One downside of this approach today is that the `kani` crate ships with Kani, but it's not available on [crates.io](https://crates.io).
+So you need to annotate the Arbitrary implementation with a `#[cfg(kani)]` attribute.
+For the derive macro, use `#[cfg_attr(kani, derive(kani::Arbitrary))]`.
+
+The second approach is recommended for cases where you would also like to be able to apply fuzzing or property testing.
```

Comment:
Another benefit over type-level trait impls is you can have a parameterized generator that can be created in the harness.

For example, `Range<T>` implements `ValueGenerator` and will then generate values that fall within those values, rather than arbitrary Range structs themselves.

```rust
bolero::check!()
    .with_generator(5..14)
    .for_each(|v| ...)
```

There's also helpers like [`one_of`](https://docs.rs/bolero-generator/0.8.0/bolero_generator/one_of/fn.one_of.html) that provide concise ways of generating specific values:

```rust
bolero::check!()
    .with_generator(one_of([123, 456, 789]))
    .for_each(|v| ...)
```

Reply:
That's fair. @danielsn suggested a variation of `any()` that takes a predicate. Which basically do a `any()` call followed by an `assume()` call which I think has a similar effect.

Do you have documentation that we can point to here?

Reply:
I don't have good documentation around generators right now... It definitely needs to be expanded.